Fix a bug in the new SH coherent mapping code (introduced in
linux-sh4-2.6.23.14_stm23_0107-coherent.patch) so that it correctly unmaps
the region when it is freed.

Signed-off-by: Stuart Menefy <stuart.menefy@st.com>
Index: linux-2.6.23-stm/arch/sh/mm/consistent.c
===================================================================
--- linux-2.6.23-stm.orig/arch/sh/mm/consistent.c
+++ linux-2.6.23-stm/arch/sh/mm/consistent.c
@@ -155,6 +155,7 @@ static struct page *__consistent_unmap(v
 
 	spin_lock_irqsave(&consistent_lock, flags);
 	c = vm_region_find(&consistent_head, (unsigned long)vaddr);
+	spin_unlock_irqrestore(&consistent_lock, flags);
 	if (!c)
 		goto no_area;
 
@@ -167,8 +168,10 @@ static struct page *__consistent_unmap(v
 
 	page = c->vm_pages;
 
-	list_del(&c->vm_list);
+	unmap_kernel_range(c->vm_start, size);
 
+	spin_lock_irqsave(&consistent_lock, flags);
+	list_del(&c->vm_list);
 	spin_unlock_irqrestore(&consistent_lock, flags);
 
 	kfree(c);
@@ -176,7 +179,6 @@ static struct page *__consistent_unmap(v
 	return page;
 
 no_area:
-	spin_unlock_irqrestore(&consistent_lock, flags);
 	printk(KERN_ERR "%s: trying to free invalid coherent area: %p\n",
 	       __func__, vaddr);
 	dump_stack();
